### PR TITLE
materialized: use structopt for main binary

### LINKED ISSUE #4191: Allow CLI flags to be passed as environment variables as well
It would be nice to be able to pass some/all of the CLI flags as environment variables, to make it easier to manage deployments in different environments. It's easy to work around it right now, but it would be more useful as more flags (like #1987) are added.

### PR DESCRIPTION
**Pull request by benesch**
Convert the materialized binary to use structopt. This is a follow-up to
PR #5298, which did the same for all of our utility binaries. The idea
is to have just one option parsing solution (i.e., structopt) throughout
our codebase.

The big benefit of using structopt is that it has built-in support for
reading flag values out of environment variables, which fixes #4191.

@sploiselle requesting your review for the docs, though obviously feel free to look at the rest too!

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5441)

### CODE REVIEW

File: doc/user/content/cli/_index.md [hunk 1]
```diff
@@ -30,6 +30,19 @@ Flag | Default | Modifies
 `-v` | N/A | Print version and exit
 `-vv` | N/A | Print version and additional build information, and exit
 
+If a command line flag named above takes an argument, you can alternatively set
+that flag via an environment variable named after the flag. For example, to set
+the data directory, you can set the `MZ_DATA_DIRECTORY` environment variable. If
+both the environment variable and command line flag are specified, the command
+line flag takes precedence.
+
+The process for converting a flag name to an environment variable name is as
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:39]: I'd make this list into a literal `ul`

File: doc/user/content/cli/_index.md [hunk 2]
```diff
@@ -30,6 +30,19 @@ Flag | Default | Modifies
 `-v` | N/A | Print version and exit
 `-vv` | N/A | Print version and additional build information, and exit
 
+If a command line flag named above takes an argument, you can alternatively set
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:33]: remove "named above"

File: doc/user/content/cli/_index.md [hunk 3]
```diff
@@ -30,6 +30,19 @@ Flag | Default | Modifies
...
+that flag via an environment variable named after the flag. For example, to set
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:34]: move this after the conversion rules

> For example, to set the data directory, you can set the `MZ_DATA_DIRECTORY` environment variable.

File: doc/user/content/cli/_index.md [hunk 4]
```diff
@@ -30,6 +30,19 @@ Flag | Default | Modifies
...
+the data directory, you can set the `MZ_DATA_DIRECTORY` environment variable. If
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:35]: I guess I'd put this in its own `p` after everything else but before `Note that command line flags...`? After doing the other shuffles, trust your heart about where this ought to go.

**Review by sploiselle — APPROVED** [review]
small, non-world-stopping edits in the docs; I glanced at the code and my eyes glazed over. I have faith in your ability to handle parsing command line flags, though.

File: doc/user/content/cli/_index.md (hunk 3 above)
Comment (benesch) [reply doc/user/content/cli/_index.md:34]: Done.

File: doc/user/content/cli/_index.md (hunk 2 above)
Comment (benesch) [reply doc/user/content/cli/_index.md:33]: Done.

File: doc/user/content/cli/_index.md (hunk 1 above)
Comment (benesch) [reply doc/user/content/cli/_index.md:39]: Done.

File: doc/user/content/cli/_index.md (hunk 4 above)
Comment (benesch) [reply doc/user/content/cli/_index.md:35]: Done!

**Pull request merged**